dataframe/field_config.go

```diff
@@ -0,0 +1,157 @@
+}
+
+// FieldConfigFromJSON create a FieldConfig from json string
+func FieldConfigFromJSON(jsonStr string) (*FieldConfig, error) {
+	var cfg FieldConfig
+	err := json.Unmarshal([]byte(jsonStr), &cfg)
+	if err != nil {
+		return nil, err
+	}
+	return &cfg, nil
+}
+
+// SetDecimals modifies the FieldConfig's Decimals property to
+// be set to v and returns the FieldConfig. It is a convenance function
+// since the Decimals property is a pointer.
+func (fc *FieldConfig) SetDecimals(v uint16) *FieldConfig {
+	fc.Decimals = &v
+	return fc
+}
+
+// SetMin modifies the FieldConfig's Min property to
+// be set to v and returns the FieldConfig. It is a convenance function
+// since the Min property is a pointer.
+func (fc *FieldConfig) SetMin(v float64) *FieldConfig {
+	fc.Min = &v
+	return fc
+}
+
+// SetMax modifies the FieldConfig's Max property to
+// be set to v and returns the FieldConfig. It is a convenance function
+// since the Min property is a pointer.
+func (fc *FieldConfig) SetMax(v float64) *FieldConfig {
+	fc.Max = &v
+	return fc
+}
+
+// SetFilterable modifies the FieldConfig's Filterable property to
+// be set to b and returns the FieldConfig. It is a convenance function
+// since the Filterable property is a pointer.
+func (fc *FieldConfig) SetFilterable(b bool) *FieldConfig {
+	fc.Filterable = &b
+	return fc
+}
+
+// NullValueMode say how the UI should show null values
+type NullValueMode string
+
+const (
+	// NullValueModeNull displays null values
+	NullValueModeNull NullValueMode = "null"
+	// NullValueModeIgnore sets the display to ignore null values
+	NullValueModeIgnore NullValueMode = "connected"
+	// NullValueModeAsZero set the display show null values as zero
+	NullValueModeAsZero NullValueMode = "null as zero"
+)
+
+// MappingType value or range
+type MappingType int8
+
+const (
+	// ValueToText map a value to text
+	ValueToText MappingType = iota + 1
+
+	// RangeToText map a range to text
+	RangeToText
+)
+
+// ValueMapping convert input value to something else
+type ValueMapping struct {
+	ID       int16       `json:"id"`
+	Operator string      `json:"operator"`
+	Text     string      `json:"title"`
+	Type     MappingType `json:"type"`
+
+	// Only valid for MappingType == ValueMap
+	Value string `json:"value,omitempty"`
+
+	// Only valid for MappingType == RangeMap
+	From string `json:"from,omitempty"`
+	To   string `json:"to,omitempty"`
+}
+
+// DataLink define what
+type DataLink struct {
+	Title       string `json:"title,omitempty"`
+	TargetBlank bool   `json:"targetBlank,omitempty"`
+	URL         string `json:"url,omitempty"`
+}
+
+// ThresholdsConfig setup thresholds
+type ThresholdsConfig struct {
+	Mode ThresholdsMode `json:"mode"`
+
+	// Must be sorted by 'value', first value is always -Infinity
+	Steps []Threshold `json:"steps"`
+}
+
+// Threshold a single step on the threshold list
+type Threshold struct {
+	Value *float64 `json:"min,omitempty"` // First value is always -Infinity serialize to null
```

Comment:
@ryantxu  I'm not clear if the first value *must* be null/-Infinity (if we do some clever encoding) *or* if the the value of `ThresholdsConfig`'s `Steps` will always be ignored (assumed -Inf). 

Besides this, I'm a bit curious why that would be? Is that limiting if you want inverse thresholds?

Reply:
First value will be overwritten -- maybe we can just use float64 directly?